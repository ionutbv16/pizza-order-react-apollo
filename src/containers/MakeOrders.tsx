
import React, { useState, useEffect } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { RouteComponentProps } from 'react-router-dom';
import gql from 'graphql-tag';
import styled from 'styled-components';
import { useForm } from "react-hook-form";
import TextField from '@material-ui/core/TextField';
import { makeStyles } from '@material-ui/core/styles';
import InputLabel from '@material-ui/core/InputLabel';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import { Button } from '@material-ui/core';
import Card from '@material-ui/core/Card';
import { IOrders } from '../types';
import { Title, Wrapper } from '../commonstyled/styled';

const CardsWrapper = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
`;

const ItemDivider = styled.div`
  padding: 10px
`;

const ButtonDivider = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    padding-top: 20px;
    padding-bottom: 30px;
`;

const ErrorMessage = styled.div`
  color: red;
  font-size: 11px;
  padding: 4px;
  width: 200px;
`;

const MUTATION_ADD_ITEM_TO_ORDERS = gql`
  mutation ($order: Any!) {
    addItemToOrders(order: $order) @client
  }
`

const useStyles = makeStyles((theme) => ({
  formControl: {
    minWidth: 194,
  },
  root: {
    minWidth: 275,
  },
  selectEmpty: {
    marginTop: theme.spacing(0),
  },
}));

const MakeOrders: React.FC<RouteComponentProps> = ({ history }) => {

  const classes = useStyles();

  const { register, handleSubmit, errors } = useForm();

  const [formorder, setFormOrder] = useState<IOrders[]>([]);

  const onSubmit = (data:IOrders) => {
    setFormOrder([data])
  };

  const [ addItemToOrders ] = useMutation(
    MUTATION_ADD_ITEM_TO_ORDERS,
    { variables: { order: formorder } }
  )

  useEffect(() => {
    if (formorder.length > 0) {
      addItemToOrders()
      history.push("/")
    }
     
  }, [formorder]);

  return (
    <Wrapper>  
      <Card className={classes.root} variant="outlined">
      <Title>Make order</Title>
      <CardsWrapper>
        <form onSubmit={handleSubmit(onSubmit)}>
            <ItemDivider>
                <TextField  id="name"  error={false} label="Name" variant="outlined"  name="name" inputRef={register({
                  validate: value => value.length > 0
                })}  />
                {errors.name && <ErrorMessage>Name is required</ErrorMessage>}
            </ItemDivider>
            <ItemDivider>
              <TextField  id="surname"  error={false} label="Surname" variant="outlined"  name="surname" inputRef={register({
                  validate: value => value.length > 0
                } ) }  />
              {errors.surname && <ErrorMessage>Surname is required</ErrorMessage>}
            </ItemDivider>
            <ItemDivider>
              <TextField  id="email"  error={false} label="Email" variant="outlined"  name="email" inputRef={register({required: true, pattern: /^\S+@\S+$/i})}  />
              {errors.email && <ErrorMessage>Invalid email format. name@domain.ext</ErrorMessage>}
            </ItemDivider>
            <ItemDivider>
              <TextField  id="telephone"  error={false} label="Telephone" variant="outlined"  name="telephone" inputRef={register({ required: true,  maxLength: 12, minLength: 6, pattern: /^\d+$/ })}  /> 
                {errors.telephone && <ErrorMessage>Invalid Telephone. Digits only (07887887887)</ErrorMessage>}
                </ItemDivider>
            <ItemDivider>
                  <FormControl variant="outlined" className={classes.formControl}>
                      <InputLabel htmlFor="pizza">Name of Pizza</InputLabel>
                      <Select
                      inputRef={register({
                        validate: value => value.length > 1
                      } ) }
                        native
                        inputProps={{
                          name: 'pizza',
                          id: 'pizza',
                        }}
                      >
                        <option aria-label="None" value="" />
                        <option value="Pepperoni">Pepperoni</option>
                        <option value="Margherita">Margherita</option>
                        <option value="Hawaiian">Hawaiian</option>
                      </Select>
                </FormControl>
                {errors.pizza && <ErrorMessage>Please select Pizza</ErrorMessage>}
            </ItemDivider>
            <ButtonDivider>
              <Button type="submit" variant="contained">Submit</Button>
            </ButtonDivider>
        </form>
      </CardsWrapper>
    </Card>  
    <ButtonDivider>
      <Button variant="contained" onClick={() => history.push("/")}>Back To Home</Button>
    </ButtonDivider>
  </Wrapper>
  );
};

export default MakeOrders;
