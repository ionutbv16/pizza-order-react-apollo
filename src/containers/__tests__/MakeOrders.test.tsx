import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/react-testing';
import wait from 'waait';
import MakeOrders from '../MakeOrders'; 

describe('Test MakeOrders Page', () => {
it('should render MakeOrders without error', async () => {
  await act(async () => {
    const component = renderer.create(
      <MockedProvider addTypename={false}>
          <MakeOrders />
      </MockedProvider>,
    );
    await wait(0);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
});
