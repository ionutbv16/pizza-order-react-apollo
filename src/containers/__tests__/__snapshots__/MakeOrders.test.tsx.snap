// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MakeOrders Page should render MakeOrders without error 1`] = `
<div
  className="sc-bwzfXH fMjrbc"
>
  <div
    className="MuiPaper-root MuiCard-root makeStyles-root-2 MuiPaper-outlined MuiPaper-rounded"
  >
    <div
      className="sc-bdVaJa cRofRy"
    >
      Make order
    </div>
    <div
      className="sc-htpNat iXDEJs"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="sc-bxivhb jAykvH"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink={false}
              htmlFor="name"
              id="name-label"
            >
              Name
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="name"
                name="name"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-6"
                >
                  <span>
                    Name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="sc-bxivhb jAykvH"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink={false}
              htmlFor="surname"
              id="surname-label"
            >
              Surname
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="surname"
                name="surname"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-6"
                >
                  <span>
                    Surname
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="sc-bxivhb jAykvH"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink={false}
              htmlFor="email"
              id="email-label"
            >
              Email
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="email"
                name="email"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-6"
                >
                  <span>
                    Email
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="sc-bxivhb jAykvH"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink={false}
              htmlFor="telephone"
              id="telephone-label"
            >
              Telephone
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="telephone"
                name="telephone"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-6"
                >
                  <span>
                    Telephone
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="sc-bxivhb jAykvH"
        >
          <div
            className="MuiFormControl-root makeStyles-formControl-1"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink={false}
              htmlFor="pizza"
            >
              Name of Pizza
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
              onClick={[Function]}
            >
              <select
                aria-invalid={false}
                className="MuiSelect-root MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="pizza"
                multiple={false}
                name="pizza"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
              >
                <option
                  aria-label="None"
                  value=""
                />
                <option
                  value="Pepperoni"
                >
                  Pepperoni
                </option>
                <option
                  value="Margherita"
                >
                  Margherita
                </option>
                <option
                  value="Hawaiian"
                >
                  Hawaiian
                </option>
              </select>
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <legend
                  className="PrivateNotchedOutline-legend-5"
                  style={
                    Object {
                      "width": 0.01,
                    }
                  }
                >
                  <span
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="sc-ifAKCX goqgDW"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Submit
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
  <div
    className="sc-ifAKCX goqgDW"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Back To Home
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
